common/src/java/org/apache/hadoop/hive/conf/HiveConf.java

```diff
@@ -4358,6 +4358,40 @@ private static void populateLlapDaemonVarsSet(Set<String> llapDaemonVarsSetLocal
       "The listener which is called when new Llap Daemon statistics is received on AM side.\n" +
       "The listener should implement the " +
       "org.apache.hadoop.hive.llap.tezplugins.metrics.LlapMetricsListener interface."),
+    LLAP_TASK_SCHEDULER_BLACKLISTING_METRICS_LISTENER_MIN_SERVED_TASKS(
+      "hive.llap.task.scheduler.blacklisting.metrics.listener.min.served.tasks", 2000,
+      "If the number of tasks served by a node is below this number then we will ignore the node\n" +
+      "when calculating the status of the cluster.\n" +
+      "Only used if hive.llap.task.scheduler.am.collect.daemon.metrics.listener is set to\n" +
+      "org.apache.hadoop.hive.llap.tezplugins.metrics.BlacklistingLlapMetricsListener"),
+    LLAP_TASK_SCHEDULER_BLACKLISTING_METRICS_LISTENER_MIN_CHANGE_DELAY(
+      "hive.llap.task.scheduler.blacklisting.metrics.listener.min.change.delay", "300s",
+      new TimeValidator(TimeUnit.SECONDS),
+      "The minimum time which should elapse between blacklisting nodes, in seconds.\n" +
+      "Only used if hive.llap.task.scheduler.am.collect.daemon.metrics.listener is set to\n" +
+      "org.apache.hadoop.hive.llap.tezplugins.metrics.BlacklistingLlapMetricsListener"),
+    LLAP_TASK_SCHEDULER_BLACKLISTING_METRICS_LISTENER_TIME_THRESHOLD(
+      "hive.llap.task.scheduler.blacklisting.metrics.listener.time.threshold", 1.5f,
+      "If the average response time of this node divided by the average response time of all the other nodes\n" +
+      "is greater than this threshold and the other conditions are satisfied too,\n" +
+      "then this node should be blacklisted.\n" +
+      "Only used if hive.llap.task.scheduler.am.collect.daemon.metrics.listener is set to\n" +
+      "org.apache.hadoop.hive.llap.tezplugins.metrics.BlacklistingLlapMetricsListener"),
+    LLAP_TASK_SCHEDULER_BLACKLISTING_METRICS_LISTENER_EMPTY_EXECUTORS(
+      "hive.llap.task.scheduler.blacklisting.metrics.listener.empty.executors.threshold", 2.0f,
```

Comment:
The default of 2x required empty executors doesn't make much sense to me. At best, a factor of 1x would be good because we still would be able to blacklist w/o any negative impact (enough empty executors available to cover). But even then, I would argue that values <1.0 make sense. If my average task execution time on a healthy node is 100ms and on a limping node is 200ms, then it is better to send the tasks (even with queueing) to the healthy nodes instead of continuing with the limping node. I would actually consider this "empty executors = free capacity" protection mechanism as solely optional and default to zero.... @t3rmin4t0r - any thoughts on this?

Reply:
My reasoning:
- The empty executor metrics is an average, so we can have peaks above this. So to be on the safe side I would aim for more available executors than we actually blacklisting
- With only 1 host generated by HostAffinitySplitLocationProvider we lose even more - Cache locality (cache locality) and rack up extra queue time as well (on Llap Daemon side definitely, and maybe even on Tez side)

Your thoughts?